src/decoder/lattice-incremental-decoder.cc

```diff
@@ -0,0 +1,1295 @@
+      const auto &arc = aiter.Value();
+      // We do not copy initial arcs, which exists except the first chunk.
+      // These arcs will be taken care later in step 3.2
+      if (!not_first_chunk || s != 0) {
+        CompactLatticeArc arc_appended(arc);
+        arc_appended.nextstate += state_offset;
+        olat->AddArc(state_appended, arc_appended);
+      }
+      // Process state labels, which will be used in step 3.2
+      if (arc.olabel > config_.max_word_id) { // initial_arc
+        if (s == 0) { // record initial_arc in this chunk, we will use it right now
+          initial_arc_map[arc.olabel] = aiter.Position();
+          if (last_get_lattice_frame_ != 0) // Erase since we are not interested
+            initial_state_in_chunk_.erase(arc.olabel);
+        } else { // final_arc
+          // record final_arc in this chunk for the step 3.2 in the next call
+          KALDI_ASSERT(clat.Final(arc.nextstate) != CompactLatticeWeight::Zero());
+          final_arc_list_.push_back(
+              pair<int32, size_t>(state_appended, aiter.Position()));
+        }
+      }
+    }
+  }
+  // sanity check, remove it later
+  // KALDI_ASSERT(initial_state_in_chunk_.size() == 0);
+
+  // step 3.2: connect the states between two chunks
+  if (last_get_lattice_frame_ != 0) {
+    KALDI_ASSERT(final_arc_list_prev_.size());
+    StateId prev_final_state = -1;
+    for (auto &i : final_arc_list_prev_) {
+      MutableArcIterator<CompactLattice> aiter(olat, i.first);
+      aiter.Seek(i.second);
+      // Obtain the appended final arcs in the last chunk
+      auto &arc_appended = aiter.Value();
+      // Find out whether its corresponding Token still exists in the begin
+      // of this chunk. If not, it is pruned by PruneActiveTokens()
+      auto r = initial_arc_map.find(arc_appended.olabel);
+      if (r != initial_arc_map.end()) {
+        ArcIterator<CompactLattice> aiter_unappended(clat, 0); // initial state
+        aiter_unappended.Seek(r->second);
+        const auto &arc_unappended = aiter_unappended.Value();
+        KALDI_ASSERT(arc_unappended.olabel == arc_appended.olabel);
+        StateId state_appended = arc_unappended.nextstate + state_offset;
+        if (prev_final_state == -1)
+          prev_final_state = arc_appended.nextstate;
+        else
+          KALDI_ASSERT(arc_appended.nextstate == prev_final_state);
+        // For the later code in this loop, we try to modify the arc_appended
+        // to connect the last frame state of last chunk to the first frame
+        // state of this chunk. These begin and final states are
+        // corresponding to the same Token, guaranteed by unique state labels.
+        CompactLatticeArc arc_appended_mod(arc_appended);
+        arc_appended_mod.nextstate = state_appended;
+
+        CompactLatticeWeight weight_offset, weight_offset_final;
+        weight_offset.SetWeight(
+            LatticeWeight(0, -state_label_initial_cost_[arc_appended.olabel]));
```

Comment:
it is bad practice to look up things in maps without checking that they are there first.  This might be masking a bug, see below.